src/state/router.js

```diff
@@ -0,0 +1,142 @@
+
+import { REQUEST_TRIGGER, API_VERSIONS_RECEIVE, REQUEST_SELECT_ENDPOINT } from './actions';
+import { defaultState as defaultRequestState } from './request/reducer';
+import { defaultState as defaultUiState } from './ui/reducer';
+import { loadEndpoints } from './endpoints/actions';
+import { getEndpoints } from './endpoints/selectors';
+
+function getUrlParams() {
+	return new URL( window.location.href ).searchParams;
+}
+
+function encodeString( val ) {
+	return ( val === null || val === undefined ) ? '' : encodeURIComponent( val );
+}
+
+function encodeObject( val ) {
+	return ( typeof val === 'object' ) && Object.keys( val ).length === 0 ? '' : encodeURIComponent( JSON.stringify( val ) );
+}
+
+function decodeString( val ) {
+	return decodeURIComponent( val );
+}
+
+function decodeObject( val ) {
+	return JSON.parse( decodeURIComponent( val ) );
+}
+
+export const getUrlFromState = state => {
+
+	const { request, ui } = state;
+
+	const params = {
+		bodyParams: encodeObject( request.bodyParams ),
+		endpoint: encodeString( request.endpoint ? request.endpoint.pathLabeled : '' ),
+		method: encodeString( request.method ),
+		pathValues: encodeObject( request.pathValues ),
+		queryParams: encodeObject( request.queryParams ),
+		url: encodeString( request.url ),
+		api: encodeString( ui.api ),
+		version: encodeString( ui.version ),
+	};
+
+	// Discard empty params
+	Object.keys( params ).forEach( key => ! params[ key ] && ( delete params[ key ] ) );
+
+	// Construct url
+	const urlParams = new URLSearchParams( params );
+	const url = window.location.origin + window.location.pathname + '?' + urlParams.toString();
+
+	return url;
+};
+
+export const getStateFromUrl = () => {
+	const urlParams = getUrlParams();
+
+	if ( urlParams.toString() === '' ) {
+		return false;
+	}
+
+	const state = { request: { ...defaultRequestState }, ui: { ...defaultUiState } };
```

Comment:
it would be best to avoid such deep coupling outside of the react tree. I can easily imagine someone adding another top-level state property and not realizing this code needs it.

I'm not sure we even need to return the full state object here as it seems like we could treat this as state augmentation. we can pass this into the initial state tree and merge it with the defaults there. that way these two parts of the system can operate independently.